### DISCUSSION TITLE
[SUGGESTION] `if (i==j)` gives clang warning about (i==j)

### DISCUSSION THREAD

**Discussion by realgdman** [hsutter/cppfront#632] [Q&A] [answered]
`if (i==j)` in cpp2, when lowered to cpp1, encloses condition in parens  `if ((i==j))`
Which is, probably, for good, but in modern C++, for example clang, gives a warning

https://cpp2.godbolt.org/z/av3ech559

>warning: equality comparison with extraneous parentheses [-Wparentheses-equality]
>note: remove extraneous parentheses around the comparison to silence this warning
>note: use '=' to turn this equality comparison into an assignment

It can be a problem if "threat warnings as error" enabled, in case of mixed cpp1/cpp2.

**Comment by JohelEGP** [comment]
Redundant parentheses are used for this:
```C++
if (a = b) // Warning, `=` instead of `==`.
if ((a = b)) // Redundant `()`s convey intention to use `=` conventionally.
```
But `if a = b` and `if (a = b)` are not valid Cpp2,
so redundant parentheses needn't be lowered to a Cpp1 _condition_ as-is.

**Comment by realgdman** [comment]
Hm, looks this cpp2 valid currently
https://cpp2.godbolt.org/z/bGPqKK3nc
```
b:= false;
c:= true;
if(b) 		{ foo(true); } else { foo(false); }
if(b=c) 	{ foo(true); } else { foo(false); }
if((b=c)) 	{ foo(true); } else { foo(false); }

i:= 0;
j:= 1;
if(i) 		{ foo(true); } else { foo(false); }
if(i=j) 	{ foo(true); } else { foo(false); }
if((i=j)) 	{ foo(true); } else { foo(false); }
```
prints 0,1,1,0,1,1

**Comment by JohelEGP** [comment, marked as answer]
You're right.
It's only `if a = b` that's rejected.
`if (a = b)` isn't.
Because the grammar is `'if' logical-or-expression`.

So only `if (a == b)` don't need the redundant parentheses lowered to Cpp1 to avoid warnings.
Or maybe it's good that they're lowered to Cpp1 as-is,
because it means, "hey, you used redundant parentheses in Cpp2!"

**Comment by JohelEGP** [comment]
> Or maybe it's good that they're lowered to Cpp1 as-is,
> because it means, "hey, you used redundant parentheses in Cpp2!"

So this is a non-bug (much less a suggestion).

**Comment by realgdman** [comment]
I should be clearer, cppfront *add* extra parens. So `if ((i==1))` becames `if (((i==1)))` etc

**Comment by JohelEGP** [comment]
It's not extra. It's the same you write.
`if a` is Cpp2, and _has_ to lower to Cpp1 `if (a)`.
